docs/CADL.md

```diff
@@ -0,0 +1,93 @@
+# Design of CADL Support
+
+## Basic Terms
+- CADL pipeline: the process and components of C# code generation from cloud services API specification written in [CADL](https://github.com/microsoft/cadl).
+- Swagger pipeline: the process and components of C# code generation from REST API specification written in [OpenAPI](https://swagger.io/).
+- generator: autorest.csharp per se, last part of both pipelines that is directly responsible for code generation.
+- generator input: file or stream with input for autorest.csharp, produced by previous part in the pipeline.
+- CADL C# Emitter: part of the CADL pipeline. Emitters are used by CADL compiler to output result of its work.
+- CodeModel.yaml: generator input in Swagger pipeline.
+- CodeModel: deserialized object representation of CodeModel.yaml in autorest.csharp.
+- cadl.json: generator input in CADL pipeline produced by CADL emitter.
+- InputModel: deserialized object representation of cadl.json in autorest.csharp.
+- OutputModel: variety of objects and object factories used by code writers to generate C# code.
+
+## CADL Pipeline Overview
+
+C# code generation CADL pipeline consists of two major components: CADL compiler that inputs API specification and outputs cadl.json using [CADL C# Emitter](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/CADL.Extension/Emitter.Csharp), and autorest.csharp (a.k.a [generator](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/AutoRest.CSharp)), executed in standalone mode, which inputs cadl.json and outputs C# files that contain  clients and models for interaction with cloud service.
+
+## CADL C# Emitter
+Emitter generates cadl.json from subset of CADL compiler output object graph. Resulting output is a standard JSON file with circular references support. It doesn't have explicit specification. Instead, it is assumed that InputModel types defined in [TS](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/CADL.Extension/Emitter.Csharp/src/type) and in [C#](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/AutoRest.CSharp/Common/Input) match each other. When new features in generator require InputModel types to be modified or extended, there is no backwards compatibility requirement. Hence, releases of autorest.csharp and CADL C# Emitter must be synchronized so that autorest.csharp can deserialize `cadl.json` produced by emitter without errors.
+
+Emitter can apply C#-specific transformations and generation rules to the spec before or in process of cadl.json generation (language-agnostic transformations and rules are expected to be applied to spec before it becomes available to the emitter).
```

Comment:
For C#-specific transformations, is there any guideline on what should be implemented in emitter or codegen?

Reply:
Not yet. I hope this will be clarified when we start putting MPG on CADL rails, cause MPG is a heavy user of such transformations.